File: web/src/dashboard/scenarioState.ts
import { explainScenario, type ExplanationReason } from "./whyModel";
import type { Scenario, Score, TestRun, Verification } from "./types";

export const DEFAULT_THRESHOLD = 80;

/** The element with the lexically greatest `id` (events are ULID-keyed → id order == time order). */
export function latestById<T extends { id: string }>(items: T[]): T | null {
  let best: T | null = null;
  for (const it of items) if (best === null || it.id > best.id) best = it;
  return best;
}

export interface ScenarioStatus {
  state: "met" | "unmet";
  latestComposite: number | null;
  latestTest: TestRun | null;
  reasons: ExplanationReason[];
}

/** Verification-aware scenario status: state/reasons from explainScenario (the canonical
 *  3-condition rule), plus latestComposite/latestTest for display. Replaces deriveScenarioState. */
export function scenarioStatus(scenario: Scenario, scores: Score[], testRuns: TestRun[], verifications: Verification[], blockedIds?: Set<string>): ScenarioStatus {
  const ex = explainScenario(scenario, scores, testRuns, verifications, blockedIds);
  const latestScore = latestById(scores.filter((s) => s.scenarioId === scenario.id));
  const latestTest = latestById(testRuns.filter((r) => r.scenarioId === scenario.id));
  return {
    state: ex.state === "met" ? "met" : "unmet",
    latestComposite: latestScore?.composite ?? null,
    latestTest,
    reasons: ex.reasons,
  };
}

/** Count how many scenarios are met (verification-aware). */
export function summarize(scenarios: Scenario[], scores: Score[], testRuns: TestRun[], verifications: Verification[], blockedIds?: Set<string>): { met: number; total: number } {
  let met = 0;
  for (const s of scenarios) if (scenarioStatus(s, scores, testRuns, verifications, blockedIds).state === "met") met++;
  return { met, total: scenarios.length };
}

File: ios/Autoloop/Domain/ScenarioState.swift
import Foundation

let DEFAULT_THRESHOLD = 80

protocol Identified { var id: String { get } }
struct IdItem: Identified { var id: String }

struct ScenarioRec: Identified { var id: String; var threshold: Int? = nil }
struct ScoreRec: Identified { var id: String; var scenarioId: String?; var composite: Double? = nil }
struct TestRunRec: Identified { var id: String; var scenarioId: String?; var failed: Int? = nil }

/// Lexically greatest id (ULID-keyed => id order == time order).
func latestById<T: Identified>(_ items: [T]) -> T? {
    items.reduce(into: T?.none) { best, it in
        if best == nil || it.id > best!.id { best = it }
    }
}

enum ScenarioMet { case met, unmet }
struct ScenarioState { let state: ScenarioMet; let latestComposite: Double?; let latestTest: TestRunRec? }

func deriveScenarioState(_ scenario: ScenarioRec, scores: [ScoreRec], testRuns: [TestRunRec]) -> ScenarioState {
    let myScores = scores.filter { $0.scenarioId == scenario.id }
    let myRuns = testRuns.filter { $0.scenarioId == scenario.id }
    let latestScore = latestById(myScores)
    let latestTest = latestById(myRuns)
    let threshold = Double(scenario.threshold ?? DEFAULT_THRESHOLD)
    let composite = latestScore?.composite
    let met = composite != nil && composite! >= threshold
        && latestTest != nil && (latestTest!.failed ?? 0) == 0
    return ScenarioState(state: met ? .met : .unmet, latestComposite: composite, latestTest: latestTest)
}

func summarize(_ scenarios: [ScenarioRec], scores: [ScoreRec], testRuns: [TestRunRec]) -> (met: Int, total: Int) {
    let met = scenarios.filter { deriveScenarioState($0, scores: scores, testRuns: testRuns).state == .met }.count
    return (met, scenarios.count)
}
